### PR TITLE
Add sitemap generator

### PR DESCRIPTION
**Pull request by vicancy**
https://github.com/dotnet/docfx/issues/1979

### CODE REVIEW

File: test/docfx.Tests/CompositeCommandTest.cs [hunk 1]
```diff
@@ -56,23 +54,37 @@ public void TestCommandFromCSCodeToHtml()
...
-        ""cwd"": ""../{Path.GetFileName(_outputFolder)}""
+        ""files"": ""api/*.yml""
     }},
-    ""dest"": ""{_outputFolder.ToNormalizedPath()}/site""
+    ""dest"": ""{_outputFolder.ToNormalizedPath()}/site"",
+    ""sitemap"":{{
+        ""baseUrl"": ""https://dotnet.github.io/docfx"",
+        ""priority"": 0.1,
+        ""changefreq"": ""monthly"",
+        ""fileOptions"":{{
```
Comment (superyyrrzz) [inline test/docfx.Tests/CompositeCommandTest.cs]: What about move to FileMappingItem?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        private IEnumerable<Tuple<string, OutputFileInfo>> GetHtmlOutputFiles(Manifest manifest)
+        {
+            if (manifest.Files == null)
+            {
+                yield break;
+            }
+
+            foreach(var file in manifest.Files)
+            {
+                if (file.DocumentType != "Toc"
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: >file.DocumentType != "Toc" [](start = 20, length = 26)

also exclude non html resources?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        private XElement GetElement(string relativePath, Uri baseUri, SitemapElementOptions options)
+        {
+            var uri = new Uri(baseUri, relativePath);
+
+            return new XElement
+                 (XName.Get("url", NamespaceName),
+                 new XElement(XName.Get("loc", NamespaceName), uri.AbsoluteUri),
+                 new XElement(XName.Get("lastmod", NamespaceName), (options.LastModified ?? DateTime.Now).ToString("yyyy-MM-ddThh:mm:ssK")),
+                 new XElement(XName.Get("changefreq", NamespaceName), (options.ChangeFrequency ?? PageChangeFrequency.Daily).ToString().ToLowerInvariant()),
+                 new XElement(XName.Get("priority", NamespaceName), options.Priority ?? 0.5)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: changefreq and priority is optional, remove default value?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                return manifest;
+            }
+
+            if (manifest.SitemapOptions.Priority.HasValue && (manifest.SitemapOptions.Priority < 0 || manifest.SitemapOptions.Priority > 1))
+            {
+                Logger.LogWarning($"Invalid priority {manifest.SitemapOptions.Priority}, priority must be between 0.0 and 1.0. Use default value 0.5 instead");
+                manifest.SitemapOptions.Priority = 0.5;
+            }
+
+            XElement urlset = new XElement(XName.Get("urlset", NamespaceName));
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]:
```cs
XNamespace ns = "http://www.sitemaps.org/schemas/sitemap/0.9";
var urlset = new XElement(ns + "urlset");
```

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                if (options.BaseUrl != manifest.SitemapOptions.BaseUrl)
+                {
+                    if (!Uri.TryCreate(options.BaseUrl, UriKind.Absolute, out currentBaseUri))
+                    {
+                        Logger.LogWarning($"Base url {options.BaseUrl} is not in a valid uri format, use base url from the default setting {manifest.SitemapOptions.BaseUrl} instead.");
+                        currentBaseUri = baseUri;
+                    }
+                }
+
+                urlset.Add(GetElement(file.Item2.RelativePath, currentBaseUri, options));
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: considering: [XStreamingElement](https://docs.microsoft.com/en-us/dotnet/api/system.xml.linq.xstreamingelement?view=netframework-4.7) + [Linq](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/how-to-perform-streaming-transform-of-large-xml-documents)

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs (hunk 4 above)
Comment (vicancy) [reply src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: Why not using `XName.Get` ?

File: test/docfx.Tests/CompositeCommandTest.cs (hunk 1 above)
Comment (vicancy) [reply test/docfx.Tests/CompositeCommandTest.cs]: use `sitemap` to make the feature options straight forward

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs (hunk 4 above)
Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: more parentheses, bad to read.

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: The next `file.OutputFiles.TryGetValue(HtmlExtension` do it

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.DocAsCode.Build.Engine/PostProcessors/SitemapGenerator.cs]: prefer to keep our own default value

**Pull request merged**